File: 1962-remove-stones-to-minimize-the-total/1962-remove-stones-to-minimize-the-total.js
/**
 * @param {number[]} piles
 * @param {number} k
 * @return {number}
 */
var minStoneSum = function(piles, k) {
  const max = Math.max(...piles);
  const cnt = new Array(max + 1).fill(0);
  let sum = 0;
  
  piles.forEach((pile) => cnt[pile]++);
  
  for (let i = max; i >= 0; i--) {
    while (cnt[i]-- > 0) {
      k-- > 0 ? cnt[Math.ceil(i / 2)]++ : sum += i;
    }
  }
  
  return sum;
};

File: 1962-remove-stones-to-minimize-the-total/1962-remove-stones-to-minimize-the-total.py
class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        mx = max(piles)
        cnt = [0] * (mx + 1)
        ans = 0
        
        for p in piles:
            cnt[p] += 1
            
        for i in range(mx, -1, -1):
            while cnt[i]:
                if k:
                    k -= 1
                    cnt[(i + 1) // 2] += 1
                else:
                    ans += i
                cnt[i] -= 1
        return ans
                    